File: ua-agent/frontend/src/shared/contracts/ping.ts
import { z } from "zod";

import { ErrorDetail, ErrorEnvelope, SCHEMA_VERSION } from "./error";

export { SCHEMA_VERSION };

// Re-exported aliases preserve the 001 import surface while sharing one envelope.
export const PingErrorDetail = ErrorDetail;
export type PingErrorDetail = z.infer<typeof PingErrorDetail>;

export const PingError = ErrorEnvelope;
export type PingError = z.infer<typeof PingError>;

export const BackendIdentity = z.object({
  name: z.literal("ua-agent"),
  version: z.string().regex(/^\d+\.\d+\.\d+(?:[-+].+)?$/, {
    message: "version must be semver",
  }),
  python: z.string().regex(/^3\.(?:11|12)\..+$/, {
    message: "python version must satisfy 3.11.* or 3.12.*",
  }),
});
export type BackendIdentity = z.infer<typeof BackendIdentity>;

export const PingSuccess = z.object({
  schema_version: z.literal(SCHEMA_VERSION),
  ok: z.literal(true),
  message: z.string().min(1).max(1024),
  echo: z.union([z.string().max(256), z.null()]),
  timestamp: z.string().datetime({ offset: false }),
  backend: BackendIdentity,
});
export type PingSuccess = z.infer<typeof PingSuccess>;

export const PingResult = z.discriminatedUnion("ok", [PingSuccess, PingError]);
export type PingResult = z.infer<typeof PingResult>;

File: ua-agent/backend/src/ua_agent/contracts/ping.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from .error import ErrorDetail, ErrorEnvelope

SCHEMA_VERSION: Literal["1"] = "1"

PingErrorDetail = ErrorDetail
PingError = ErrorEnvelope


class BackendIdentity(BaseModel):
    """Identity of the backend that produced a payload."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    name: Literal["ua-agent"] = "ua-agent"
    version: Annotated[str, Field(pattern=r"^\d+\.\d+\.\d+(?:[-+].+)?$")]
    python: Annotated[str, Field(pattern=r"^3\.(?:11|12)\..+$")]


class PingSuccess(BaseModel):
    """Successful response of the `ping` subcommand."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["1"] = SCHEMA_VERSION
    ok: Literal[True] = True
    message: Annotated[str, Field(min_length=1, max_length=1024)]
    echo: Annotated[str | None, Field(max_length=256)]
    timestamp: Annotated[
        str,
        Field(
            description=(
                "RFC 3339 UTC, ms precision, trailing Z. "
                "Example: 2026-05-02T12:34:56.789Z."
            ),
        ),
    ]
    backend: BackendIdentity


PingResult = Annotated[
    PingSuccess | ErrorEnvelope,
    Field(discriminator="ok"),
]
